### ISSUE TITLE
Optimization fails even thought start value very close to best-fit

### ISSUE THREAD

**Issue by hamogu** [sherpa/sherpa#1102]
```
from sherpa import stats, optmethods
from sherpa.models import NormGauss1D
from sherpa.astro.models import Lorentz1D
from sherpa.data import Data1DInt
from sherpa.fit import Fit

mydata = Data1DInt('mydata', [104.86674388, 104.87674388, 104.88674388, 104.89674388,
        104.90674388, 104.91674388, 104.92674388, 104.93674388,
        104.94674388],
                  [104.87674388, 104.88674388, 104.89674388, 104.90674388,
        104.91674388, 104.92674388, 104.93674388, 104.94674388,
        104.95674388],
                  [ 167.43850424,  545.00692715, 1738.72267089, 5081.0329047 ,
       9908.85202355, 9707.14990356, 4948.34854889, 1788.56162686,
        577.87318162],
                  staterror=[12.9398031 , 23.34538342, 41.69799361, 71.28136436, 99.54321686,
       98.52486947, 70.34449907, 42.29138951, 24.03899294])

g1 = NormGauss1D(name='GAUSS1')
g1.pos = 104.91
g1.fwhm = 0.01

g1.fwhm.min = 0.005
g1.fwhm.max = 0.05
g1.ampl.min = 0


l1 = Lorentz1D(name="LORENTZ1")
l1.pos = 104.91
l1.fwhm = 0.01
l1.fwhm.min = 0.005
l1.fwhm.max = 0.05
l1.ampl.min = 0

model = g1 + l1
modfit = Fit(sdata, model, stat=stats.Chi2(), method=optmethods.LevMar())
res = modfit.fit()
```

The result looks like this (sherpa 4.13, installed though conda with full CIAO installation on MacOS):
![image](https://user-images.githubusercontent.com/498688/107308157-1fe89500-6a56-11eb-8afb-b147d97898d1.png)

That's a terrible fit (NelderMead looks similar). On the other hand, Moncar does great, it just runs a little longer:
![image](https://user-images.githubusercontent.com/498688/107308306-66d68a80-6a56-11eb-81d6-7695e2a6dfed.png)

This is taken from a real-world example (Arcus line-spread functions). I have hardcoded the numbers in the example above to make sure that this issue is self-contained when someone looks at it later, but I've seen this several times, not just with these specific numbers. 

I don't know what the reasons is. Maybe it's because the datavalues span more than one order of magnitude? However, that's not crazy to fit that. So, I wonder, is there any default that can we change in LevMar and NelderMead to make this work better out-of-the-box? 

I've played with the ftol, xtol, gtol, and epsfcn values for LevMar. When I make those smaller (say 1e-14), the fit improves, but is still not at good as Moncar.

Note: I do have a solution that works for me (Moncar), but I'm a little disappointed that LevMar and NelderMead fail miserably with default values and I can't get them to work well with any setting. I'm reporting this in the hope that we can do better.

**Comment by dtnguyen2** [comment]
The Lorentz1D model function will generate values using double precision so the optimal value for ```epsfcn``` should be:

```
print(np.finfo(float).eps)
# 2.22044604925e-16
```
The default value for ```epsfcn``` is:
```
print(np.finfo(np.float32).eps)
# 1.19209e-07
```
because that is the optimal value for single precision model functions (which is what most XSPEC functions were when we initially released sherpa)

I still have some work to do for ```L3``` so can't look into this right now, but there are a bunch of fitting algorithms that are not publicly available.  After a little work (but a lot of work involved before release to the public), I managed to get it to get the same results (within tolerance) as the shipped ```MonCar``` but with  a little less time consuming.

Get the result for ```MonCar``` so we can compare fit-results and time with other non-publicly available opt methods

```
(sherpa) [dtn@devel12 minuit-scipy]$ time tst.py -o moncar > use_moncar
/export/sherpa/sherpa/data.py:50: UserWarning: Converting array [167.43850424, 545.00692715, 1738.72267089, 5081.0329047, 9908.85202355, 9707.14990356, 4948.34854889, 1788.56162686, 577.87318162] to numpy array.
  warnings.warn("Converting array {} to numpy array.".format(array))

real	0m3.764s
user	0m4.183s
sys	0m1.664s
```

Using ```MonCar``` the final ```rstat``` is:
```
(sherpa) [dtn@devel12 minuit-scipy]$ grep rstat use_moncar
rstat          = 4.155195349497589
```

However, before showing you other, not yet available, options from sherpa let's iterate through some of the optimization routines that are available from ```scipy``` to see how well they do:

```
(sherpa) [dtn@devel12 minuit-scipy]$ time tst.py -o scipy

 trust-constr  method took  0.46712255477905273 secs
rstat = 11480.361330997393

 nelder-mead  method took  0.08845853805541992 secs
/export/miniconda/envs/sherpa/lib/python3.7/site-packages/scipy/optimize/_minimize.py:522: RuntimeWarning: Method nelder-mead cannot handle constraints nor bounds.
  RuntimeWarning)
rstat = 11485.969152484751

 powell  method took  0.6989805698394775 secs
rstat = 4.155199942130088

 cg  method took  0.10607600212097168 secs
rstat = 1109.703849978556

 bfgs  method took  0.041559457778930664 secs
rstat = 6716.939437702688

 l-bfgs-b  method took  0.009273529052734375 secs
rstat = 7523.276589847694

 tnc  method took  0.06882667541503906 secs
rstat = 11487.662096973692

 cobyla  method took  0.0030775070190429688 secs
rstat = 11485.756292002041

 slsqp  method took  0.0073621273040771484 secs
rstat = 11484.200183076866

```
It looks like only the method ```powell``` from ```scipy``` is the only one that gets the result that you are satisfy with.  I _think_ the 'l-bfgs-b' method is the default algorithm on scipy.  BTW, the implementation of the nelder-mead algorithm in ```scipy``` does not handle bounds.

How about ```minuit```?

```
(sherpa) [dtn@devel12 minuit-scipy]$ time tst.py -o minuit > use_minuit
/export/sherpa/sherpa/data.py:50: UserWarning: Converting array [167.43850424, 545.00692715, 1738.72267089, 5081.0329047, 9908.85202355, 9707.14990356, 4948.34854889, 1788.56162686, 577.87318162] to numpy array.
  warnings.warn("Converting array {} to numpy array.".format(array))

real	0m1.262s
user	0m1.628s
sys	0m1.721s

(sherpa) [dtn@devel12 minuit-scipy]$ diff use_moncar use_minuit 

<  moncar  method took  2.7429986000061035 secs
---
>  minuit  method took  0.1940598487854004 secs
10c1028
< methodname     = moncar
---
> methodname     = myminuit
12c1030
< succeeded      = True
---
> succeeded      = 0
14,15c1032,1033
< parvals        = (0.031817590371712806, 104.91691467882073, 24074.108539901485, 0.016491557137575805, 104.91621229783516, 11790.535128692574)
< statval        = 12.465586048492767
---
> parvals        = (-3.687223749824619e+25, 104.92566145030646, 15404.40955215696, -7.270181463379594e+31, 104.9078628591277, 16399.503902386314)
> statval        = 3503.381600594767
17c1035
< dstatval       = 34446.66199289156
---
> dstatval       = 30955.745978345287
20,23c1038,1041
< qval           = 0.005947112360864856
< rstat          = 4.155195349497589
< message        = successful termination
< nfev           = 9933
---
> qval           = 0.0
> rstat          = 1167.7938668649224
> message        = Optimization terminated successfully.
> nfev           = 666
```

```minuit``` can get the correct answer if the optimization routine is called without bounds.  I suspect that is because internally ```minuit``` handles the simple bounds by transforming and because the bounds as set are too wide.  The caveat from the ```minuit``` manual is:

>For variable parameters with limits, MINUIT uses the following transformation:

>  P   = arcsin(2((P   -a)/(b- a))-1)                P    = a+((b- a)/(2))(sinP    +1)
>   int             ext                               ext                      int
> so that the internal value P can take on any value, while the external int value P can take on values only between the lower limit a and the ext upper limit b. Since the transformation is necessarily non-linear, it would transform a nice linear problem into a nasty non-linear one, which is the reason why limits should be avoided if not necessary. In addition, the transformation does require some computer time, so it slows down the computation a little bit, and more importantly, it introduces additional numerical inaccuracy into the problem in addition to what is introduced in the numerical calculation of the FCN value. The effects of non-linearity and numerical roundoff both become more important as the external value gets closer to one of the limits (expressed as the distance to nearest limit divided by distance between limits). The user must therefore be aware of the fact that, for example, if he puts limits of (0,10^10 ) on a parameter, then the values 0.0 and 1. 0 will be indistinguishable to the accuracy of most machines.

```
(sherpa) [dtn@devel12 minuit-scipy]$ time tst.py -o minuit > use_minuit_nobounds
/export/sherpa/sherpa/data.py:50: UserWarning: Converting array [167.43850424, 545.00692715, 1738.72267089, 5081.0329047, 9908.85202355, 9707.14990356, 4948.34854889, 1788.56162686, 577.87318162] to numpy array.
  warnings.warn("Converting array {} to numpy array.".format(array))

real	0m1.121s
user	0m1.510s
sys	0m1.701s
(sherpa) [dtn@devel12 minuit-scipy]$ diff use_moncar use_minuit_nobounds 
6c23
<  moncar  method took  2.7429986000061035 secs
---
>  minuit  method took  0.07165670394897461 secs
10c27
< methodname     = moncar
---
> methodname     = myminuit
12c29
< succeeded      = True
---
> succeeded      = 1
14,15c31,32
< parvals        = (0.031817590371712806, 104.91691467882073, 24074.108539901485, 0.016491557137575805, 104.91621229783516, 11790.535128692574)
< statval        = 12.465586048492767
---
> parvals        = (0.03181797310436759, 104.91691448401681, 24074.5595839431, 0.016491078388265485, 104.91621273287744, 11789.655728182695)
> statval        = 12.46559283744384
17c34
< dstatval       = 34446.66199289156
---
> dstatval       = 34446.66198610261
20,23c37,40
< qval           = 0.005947112360864856
< rstat          = 4.155195349497589
< message        = successful termination
< nfev           = 9933
---
> qval           = 0.005947093580639484
> rstat          = 4.15519761248128
> message        = Optimization terminated successfully.
> nfev           = 466
```

So calling ```minuit``` without bounds gets the correct answer. I have suggested that we ship ```minuit``` with sherpa but for some reason SDS has said no.

Using ```MonCar``` but with ```LevMar``` for local optimization algorithm:

```
(sherpa) [dtn@devel12 minuit-scipy]$ time tst.py -o moncarlevmar > use_moncarlevmar
/export/sherpa/sherpa/data.py:50: UserWarning: Converting array [167.43850424, 545.00692715, 1738.72267089, 5081.0329047, 9908.85202355, 9707.14990356, 4948.34854889, 1788.56162686, 577.87318162] to numpy array.
  warnings.warn("Converting array {} to numpy array.".format(array))

real	0m1.238s
user	0m1.648s
sys	0m1.698s
```

Using a hacked version of ```NelderMead```:
```
(sherpa) [dtn@devel12 minuit-scipy]$ time tst.py -o ncoresneldermead > use_ncores_neldermead
/export/sherpa/sherpa/data.py:50: UserWarning: Converting array [167.43850424, 545.00692715, 1738.72267089, 5081.0329047, 9908.85202355, 9707.14990356, 4948.34854889, 1788.56162686, 577.87318162] to numpy array.
  warnings.warn("Converting array {} to numpy array.".format(array))
[1]+  Terminated              tst.py -o numcoresneldermead > use_numcores_neldermead

real	0m1.628s
user	0m2.437s
sys	0m1.784s
```

Compare the shipped ```MonCar``` with ```MonCarLevMar```.  Similar results, but about 10 times faster.
```
(sherpa) [dtn@devel12 sherpa]$ diff use_moncar use_moncar_levmar
6c6
<  moncar  method took  2.7429986000061035 secs
---
>  moncarlevmar  method took  0.19718313217163086 secs
10c10
< methodname     = moncar
---
> methodname     = moncarlevmar
14,15c14,15
< parvals        = (0.031817590371712806, 104.91691467882073, 24074.108539901485, 0.016491557137575805, 104.91621229783516, 11790.535128692574)
< statval        = 12.465586048492767
---
> parvals        = (0.03181709971091819, 104.9169147245778, 24073.668926913895, 0.016493293729917268, 104.91621218661373, 11791.121789685802)
> statval        = 12.465589589311286
17c17
< dstatval       = 34446.66199289156
---
> dstatval       = 34446.661989350745
20,23c20,23
< qval           = 0.005947112360864856
< rstat          = 4.155195349497589
< message        = successful termination
< nfev           = 9933
---
> qval           = 0.005947102565917338
> rstat          = 4.1551965297704285
> message        = Optimization terminated successfully
> nfev           = 942
```

Comparing ```MonCar``` with a hacked ```NelderMead```. Similar result but about 4 times faster
```
6c6
<  moncar  method took  2.7429986000061035 secs
---
>  moncarlevmar  method took  0.19718313217163086 secs
10c10
< methodname     = moncar
---
> methodname     = moncarlevmar
14,15c14,15
< parvals        = (0.031817590371712806, 104.91691467882073, 24074.108539901485, 0.016491557137575805, 104.91621229783516, 11790.535128692574)
< statval        = 12.465586048492767
---
> parvals        = (0.03181709971091819, 104.9169147245778, 24073.668926913895, 0.016493293729917268, 104.91621218661373, 11791.121789685802)
> statval        = 12.465589589311286
17c17
< dstatval       = 34446.66199289156
---
> dstatval       = 34446.661989350745
20,23c20,23
< qval           = 0.005947112360864856
< rstat          = 4.155195349497589
< message        = successful termination
< nfev           = 9933
---
> qval           = 0.005947102565917338
> rstat          = 4.1551965297704285
> message        = Optimization terminated successfully
> nfev           = 942
```

**Comment by anetasie** [comment]
@dtnguyen2 these are good method comparison tests we can include in our documentation.

**Comment by anetasie** [comment]
@dtnguyen2 How to access 'moncarlevmar' option?

**Comment by dtnguyen2** [comment]
One should not make a decision on which optimization algorithm, or more precisely which implementation of an algorithm (e.g. scipy's implementation of the Nelder-Mead's algorithm differ from Sherpa's), to use on the basis of one problem.  In layman's term: It is impossible for one implementation of an optimization algorithm to minimize all problems, a more precise way of saying can be found by doing a web search for "no free lunch theorem of optimization" (personally, I always think of it as the full employment act :).

As for ```MonCar``` with ```LevMar``` there is still a lot of work to be done before a public release.

**Comment by hamogu** [comment]
@dtnguyen2 Of course you are right in principle, but in practice, sherpa is mostly used in the X-ray astronomy, where the range of problems is not actually not that wide - X-ray spectra all look pretty similar. So maybe the specific example in here this thread is not the most typical, but it's probably the second most typical.

So, just 2-3 problems (e.g. X-ray powerlaw, fit a single line, fit grating spectrum) cover probably > 75% all current sherpa use cases, so documentation that discusses just one or two (with the caveat that you raised of source!) is going to be valuable for a large fraction of our users.
